### DISCUSSION TITLE
Possible to add to constructor from property aspects?

### DISCUSSION THREAD

**Discussion by etkramer** [pamidur/aspect-injector#194] [General]
My goal is to give `NotifyAttribute` a parameter "dependsOn", which causes it to be notified any time some other property is changed - basically the reverse of AlsoNotify. To do this, my first instinct was to add something along the lines of
```cs
[Advice(Kind.After, Targets = Target.Constructor)]
public void AfterConstructor([Argument(Source.Instance)] object source, [Argument(Source.Name)] string propertyName, [Argument(Source.Triggers)] Attribute[] triggers)
{
    foreach (string dependencyProperty in triggers[0].DependsOn)
    {
        source.Subscribe(dependencyProperty, () => source.Raise(propertyName));
    }
}
```
However, this Advise doesn't actually seem to be triggered unless the aspect is applied directly to the class or the constructor - is there any way to have this be run used when the attribute is applied to a _property_ instead?

**Comment by pamidur** [comment]
Hi @BitCrusader ,thanks for your interest in the project. You have an interesting idea here. 

Let's consider this class 
```c#
class Test
{
  [Notify]
  public string Name {get; set;}

  [Notify (DependsOn = nameof(Name) )]
  public string FirstName {get; set;}
}
```

In this case ``` [Notify (DependsOn = nameof(Name) )]``` will only be triggered if ```FirstName ``` is changed, but will not be triggered if ```Name ``` is changed. As per [docs](https://github.com/pamidur/aspect-injector/blob/master/docs/injection.md#injection-propagation)

When you apply 
```c#
[Notify (DependsOn = nameof(Name) )]
class Test
{
  [Notify]
  public string Name {get; set;}

  [Notify (DependsOn = nameof(Name) )]
  public string FirstName {get; set;}
}
```
you effectively doing this:
```c#
class Test
{
  [Notify]
  [Notify (DependsOn = nameof(Name) )]
  public string Name {get; set;}

  [Notify (DependsOn = nameof(Name) )]
  [Notify (DependsOn = nameof(Name) )]
  public string FirstName {get; set;}
}
```

but it still won't trigger Aspect on ```FirstName ``` if ```Name``` is changed. But  you might see behaviour where :
```c#
  [Notify (DependsOn = nameof(Name) )]
  public string Name {get; set;}
```


To solve it you might need  to keep track on the dependencies in the class, but figure this dependencies on the first hit. Like the following:
- Advice is executed for ```Name```
- You get all attributes of all properties of the class via reflection and figure which ones have ```DependsOn```. Find ```FirstName```
- Cache optionally
- Raise events for dependent properties

**Comment by pamidur** [comment]
If it already the case. And you just need to apply aspects to all properties, you might just want to have different attributes. Like this:
```c#
[Notifiable]
class Test
{
  [Notify]
  public string Name {get; set;}

  [Notify (DependsOn = nameof(Name) )]
  public string FirstName {get; set;}

  public string FirstName {get; set;}
}
```

so like this Aspect will be executed for all props and then you can raise events only for those having ```[Notify]```
